bulkimport/src/main/java/com/microsoft/azure/documentdb/bulkimport/DocumentBulkImporter.java

```diff
@@ -554,7 +582,138 @@ private int getDocumentSizeOrThrow(String document) {
 
         return futureContainer.callAsync(completeAsyncCallback, listeningExecutorService);
     }
+    
+    private ListenableFuture<BulkUpdateResponse> executeBulkUpdateAsyncImpl(Collection<UpdateItem> updateItems) {
+        Stopwatch watch = Stopwatch.createStarted();
+
+        logger.debug("Bucketing update items ...");
+
+        ConcurrentHashMap<String, Set<UpdateItem>> updateItemsByPartition = new ConcurrentHashMap<String, Set<UpdateItem>>();
+        ConcurrentHashMap<String, List<List<UpdateItem>>> miniBatchesToUpdateByPartition = new ConcurrentHashMap<String, List<List<UpdateItem>>>();
+
+        for (String partitionKeyRangeId: partitionKeyRangeIds) {
+        	updateItemsByPartition.put(partitionKeyRangeId,  ConcurrentHashMap.newKeySet(updateItems.size() / partitionKeyRangeIds.size()));
+        	miniBatchesToUpdateByPartition.put(partitionKeyRangeId, new ArrayList<List<UpdateItem>>(1000));
+        }
+
+        updateItems.parallelStream().forEach(updateItem -> {
+            PartitionKeyInternal partitionKeyValue = DocumentAnalyzer.fromPartitionKeyvalue(updateItem.getPartitionKeyValue());
+            String effectivePartitionKey = partitionKeyValue.getEffectivePartitionKeyString(partitionKeyDefinition, true);
+            String partitionRangeId = collectionRoutingMap.getRangeByEffectivePartitionKey(effectivePartitionKey).getId();
+            updateItemsByPartition.get(partitionRangeId).add(updateItem);
+        });
 
+        logger.trace("Creating mini batches within each partition bucket");
+
+        updateItemsByPartition.entrySet().parallelStream().forEach(entry -> {
+
+            String partitionRangeId = entry.getKey();
+            Set<UpdateItem> updateItemsInPartition =  entry.getValue();
+
+            Iterator<UpdateItem> it = updateItemsInPartition.iterator();
+            ArrayList<UpdateItem> currentMiniBatch = new ArrayList<UpdateItem>(500);
+            int currentMiniBatchIndex = 0;
+
+            while (it.hasNext()) {
+                UpdateItem currentUpdateItem = it.next();
+
+                if ((currentMiniBatchIndex + 1 <= 10000)) {
```

Comment:
what's the significant of "10000"?

Reply:
It is the max count of update item objects in the payload to the sproc. Because Java doesn't have a sizeof equivalent, forced to resort to count and not size. Moved this to taking it in through the builder now